### PR TITLE
print haplotypes by rows

### PR DESCRIPTION
**Pull request by shajoezhu**
Introducing options "-print-transpose" and "--print-transpose", to print the haplotypes in a table, where the first column is the mutation position, second column is the age of the mutation, and the rest of columns denote the hapotypes

### CODE REVIEW

**Comment by shajoezhu** [comment]
Hi @paulstaab , I was wondering could you guide me how to generate the summary table that you have in https://github.com/scrm/scrm/pull/83. I think this pull request is potentially useful for users, so they won't have to reshape the data when parsing into R.

Best
Joe

**Comment by paulstaab** [comment]
I would prefer it transposed would be an option of the `SegSites` class. So that if you initialize SegSites with `transposed = true`, it prints the transposed output on `printLocusOutput`, and the normal output otherwise.

This wouldn't need any changes in `scrm.cc` and `Forest`, and keep the code close to were it actually produces an effect.

You can produces the table by running `./tools/compare_versions.sh master transpose`. Requires git, make/autoconf & R.

**Comment by paulstaab** [comment]
Also, I'd like to add some some unit tests before we merge that in. What do you think about making it an option of `SegSites`? I totally agree that it is a useful addition.

**Comment by shajoezhu** [comment]
yes, I agree. should have stick with the frame work, and making minimal changes.

**Comment by shajoezhu** [comment]
@paulstaab what sort of cases are you think of for the unit testing?

**Comment by paulstaab** [comment]
I changed the argument to `-transpose-segsites`, that's a bit more descriptive. Hope that is okay for you.

For the tests:
- we only check the internal seg_sites structure, but not the output produced. I guess that should be sufficient.
- I added a quick check that parsing the arguments works to `test_param.cc`
- I added the argument to two runs in `test_binaries.sh` to ensure that there are no memory errors involved (not that the code looks like there could be any...)

**Comment by paulstaab** [comment]
What do you think about changing the output from

```
segsites: 3
transpose is 1
positions height 1 2 3
0.05800699 0.6803861 0 1 0
0.1223096 0.3222072 0 1 1
0.9076263 0.3189613 0 1 0
```

to 

```
transposed segsites
positions height 1 2 3
0.05800699 0.6803861 0 1 0
0.1223096 0.3222072 0 1 1
0.9076263 0.3189613 0 1 0
```
?

The information how many segsites there are is not needed for parsing this output, so I think it can be removed without loosing much. And having the first line different from what is was before should make parsers that look for the old format ignore this one.

File: tests/unittests/test_param.cc [hunk 1]
```diff
@@ -510,6 +510,16 @@ class TestParam : public CppUnit::TestCase {
     CPPUNIT_ASSERT( !model.has_window_seq() );
     CPPUNIT_ASSERT_EQUAL( (size_t)10, model.window_length_rec() );
   }
+
+  void testTransposeSegSites() {
+    Model model = Param("2 2 -r 10 100 -t 5 -print-transpose").parse();
```
Comment (shajoezhu) [inline tests/unittests/test_param.cc]: `-transpose-segsites`

**Comment by shajoezhu** [comment]
Hi Paul

Thanks for making the unittest case, and check binary case. Yes, `transpose is 1` was a debug message that I left in the code.... oops...

Yes, I agree with changing the flag to `-transpose-segsites` as well. It is more descriptive.

best
joe

File: tests/unittests/test_param.cc (hunk 1 above)
Comment (paulstaab) [inline tests/unittests/test_param.cc]: Oops, the test isn't actually called. that's why it's not failing...

**Comment by paulstaab** [comment]
Command | Equal results | time master | time transpose | memory master | memory transpose
------- | ------------- | ----------- | -------- | ------------- | ---------
scrm 100 100 -t 5 -T | Yes | 0.092s | 0.093s | 3158kb | 3185kb
scrm 20 100 -r 10 100 -t 3 -T -L | Yes | 0.089s | 0.089s | 3145kb | 3162kb
scrm 4 10000 -r 10 100 -t 5 | Yes | 0.53s | 0.55s | 3053kb | 3925kb
scrm 20 2 -r 500 100 -L | Yes | 0.15s | 0.16s | 3204kb | 3221kb
scrm 2000 1 -r 10 100 -l 50 -L -T | Yes | 0.56s | 0.56s | 5986kb | 5995kb
scrm 20 1 -r 4000 10000000 -l 300000 -T | Yes | 0.89s | 0.9s | 4408kb | 4407kb
scrm 500 1 -r 1000 10000000 -l 0 -T | Yes | 0.71s | 0.72s | 3897kb | 3901kb
scrm 100 1 -r 400 10000000 -l 0 -O | Yes | 0.3s | 0.3s | 3194kb | 3193kb
scrm 10 100 -r 10 100 -I 3 3 3 4 0.5 -eN 0.1 0.05 -eN 0.2 0.5 -L | Yes | 0.054s | 0.056s | 3119kb | 3123kb
scrm 50 100 -r 20 200 -G 1.5 -l 100 -L | Yes | 0.046s | 0.046s | 3147kb | 3146kb

**Comment by paulstaab** [comment]
From my side this is good to go :+1:. If you are happy as with it as well, fell free to do the honors ;-)
Thanks for your work here

**Comment by shajoezhu** [comment]
Hi Paul, thanks for checking! 

Best
Joe

**Pull request merged**